File: frontend/src/api/graphLayout.ts
import type { ApiResponse, GraphLayoutDataV1 } from '@campaigner/shared';
import type {
  DeleteGraphLayoutInput as TauriDeleteGraphLayoutInput,
  GetGraphLayoutInput as TauriGetGraphLayoutInput,
  GraphLayoutDataV1 as TauriGraphLayoutDataV1,
  GraphLayoutDataV1_Serialize as TauriGraphLayoutDataV1Serialize,
  GraphLayoutNodeState as TauriGraphLayoutNodeState,
  GraphLayoutResponse as TauriGraphLayoutResponse,
  GraphLayoutViewport as TauriGraphLayoutViewport,
  UpsertGraphLayoutInput_Serialize as TauriUpsertGraphLayoutInput,
} from '@/types/generated/bindings';
import { transport } from './transport';

type ApiResult<T> = {
  data: ApiResponse<T>;
};

const isApiResponse = <T>(value: unknown): value is ApiResponse<T> =>
  Boolean(value && typeof value === 'object' && 'success' in value);

const toViewport = (viewport: TauriGraphLayoutViewport) => ({
  x: viewport.x ?? 0,
  y: viewport.y ?? 0,
  zoom: viewport.zoom ?? 1,
});

const toNodeState = (node: TauriGraphLayoutNodeState) => ({
  x: node.x ?? 0,
  y: node.y ?? 0,
  pinned: node.pinned ?? undefined,
});

const toGraphLayoutData = (layout: TauriGraphLayoutDataV1): GraphLayoutDataV1 => ({
  version: 1,
  viewport: layout.viewport ? toViewport(layout.viewport) : undefined,
  nodes: Object.fromEntries(
    Object.entries(layout.nodes ?? {}).map(([id, node]) => [id, toNodeState(node)]),
  ),
});

const toGraphLayoutResponse = (
  response: ApiResponse<{ layoutData: GraphLayoutDataV1 }> | TauriGraphLayoutResponse,
): ApiResult<{ layoutData: GraphLayoutDataV1 }> => {
  if (isApiResponse<{ layoutData: GraphLayoutDataV1 }>(response)) {
    return { data: response };
  }

  return {
    data: {
      success: true,
      data: {
        layoutData: toGraphLayoutData(response.layoutData),
      },
    },
  };
};

const toTauriLayoutData = (layout: GraphLayoutDataV1): TauriGraphLayoutDataV1Serialize => ({
  version: layout.version,
  viewport: layout.viewport
    ? {
        x: layout.viewport.x,
        y: layout.viewport.y,
        zoom: layout.viewport.zoom,
      }
    : null,
  nodes: Object.fromEntries(
    Object.entries(layout.nodes).map(([id, node]) => [
      id,
      {
        x: node.x,
        y: node.y,
        pinned: node.pinned ?? null,
      },
    ]),
  ),
});

export const graphLayoutApi = {
  get: async (
    projectId: number,
    params: { graphType: string; branchId?: number },
  ): Promise<ApiResult<{ layoutData: GraphLayoutDataV1 }>> => {
    const input: TauriGetGraphLayoutInput = {
      projectId,
      graphType: params.graphType,
      branchId: params.branchId ?? null,
    };

    const response = await transport.request<
      ApiResponse<{ layoutData: GraphLayoutDataV1 }> | TauriGraphLayoutResponse
    >({
      http: {
        method: 'GET',
        path: `/projects/${projectId}/graph-layout`,
        query: params,
      },
      tauri: {
        command: 'graph_layout_get',
        args: { input },
      },
    });

    return toGraphLayoutResponse(response);
  },

  put: async (
    projectId: number,
    body: { graphType: string; layoutData: GraphLayoutDataV1; branchId?: number },
  ): Promise<ApiResult<{ layoutData: GraphLayoutDataV1 }>> => {
    const input: TauriUpsertGraphLayoutInput = {
      projectId,
      graphType: body.graphType,
      layoutData: toTauriLayoutData(body.layoutData),
      branchId: body.branchId ?? null,
    };

    const response = await transport.request<
      ApiResponse<{ layoutData: GraphLayoutDataV1 }> | TauriGraphLayoutResponse
    >({
      http: {
        method: 'PUT',
        path: `/projects/${projectId}/graph-layout`,
        body,
      },
      tauri: {
        command: 'graph_layout_upsert',
        args: { input },
      },
    });

    return toGraphLayoutResponse(response);
  },

  delete: async (
    projectId: number,
    params: { graphType: string; branchId?: number },
  ): Promise<{ data: void }> => {
    const input: TauriDeleteGraphLayoutInput = {
      projectId,
      graphType: params.graphType,
      branchId: params.branchId ?? null,
    };

    await transport.request<void>({
      http: {
        method: 'DELETE',
        path: `/projects/${projectId}/graph-layout`,
        query: params,
      },
      tauri: {
        command: 'graph_layout_delete',
        args: { input },
      },
    });

    return { data: undefined as void };
  },
};

File: src-tauri/src/models/graph_layout.rs
use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use specta::Type;

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct GraphLayoutViewport {
    pub x: f64,
    pub y: f64,
    pub zoom: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct GraphLayoutNodeState {
    pub x: f64,
    pub y: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pinned: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct GraphLayoutDataV1 {
    pub version: i32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub viewport: Option<GraphLayoutViewport>,
    #[serde(default)]
    pub nodes: HashMap<String, GraphLayoutNodeState>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct GetGraphLayoutInput {
    pub project_id: i32,
    pub graph_type: String,
    pub branch_id: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct UpsertGraphLayoutInput {
    pub project_id: i32,
    pub graph_type: String,
    pub layout_data: GraphLayoutDataV1,
    pub branch_id: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct DeleteGraphLayoutInput {
    pub project_id: i32,
    pub graph_type: String,
    pub branch_id: Option<i32>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct GraphLayoutResponse {
    pub layout_data: GraphLayoutDataV1,
}
